C/C++ frontends: select standard using the same rules as the compiler we emulate

smowton opened the pull request:
Both GCC and Clang have changed the C and C++ standards they pick by default when no `-std` argument is passed. Since the preprocessor is invoked also with no `-std` either, it may define e.g. `__cplusplus` or `__STDC_VERSION__` in a way that causes standard library includes to use C++14 or C11 features -- in this case `goto-cc` needs to expect that and parse them correctly.

This PR also amends our Travis config to symlink the compiler we build with as `$HOME/bin/gcc` and `$HOME/bin/g++` (even if it's really Clang, since Clang-named-gcc is common e.g. on OSX), so that our `goto-cc` tests find this version as their underlying compiler (the one it uses for preprocessing). Previously `goto-cc` C tests were using `gcc-5` (even when built with Clang) but C++ tests were using `g++-4.8`, the system default on Ubuntu 14.04.

Finally I add builds using `gcc-8` and `clang-6` (the latest version of each compiler). I can add complete testing for GCC {5, 6, 7, 8} and Clang {3.7, 3.8, 3.9, 4, 5, 6}, but I figured that might run up too large of a bill.

smowton committed:
goto-gcc: select default language standard depending on emulated compiler

This ensures that goto-cc with no `-std` explicitly specified will expect the same
language standard as the underlying compiler's preprocessor -- so for example, using
`goto-cc test.cpp`, which invokes `g++-7 -E` to perform preprocessing, will be able to
parse the C++14 constructs in the standard library exposed because g++-6 and above
enable C++14 support by default.

smowton committed:
Amend two tests that explicitly target C++98

These were previously broken on g++-6 and above, as the compiler defaults to C++14 mode.

smowton committed:
Travis: test more g++ and clang versions

Also use the build compiler, not the system default, for tests. This was already done
with gcc (by placing a symlink named 'gcc' pointing to gcc-5 in a directory ahead of /usr/bin
in the path), but g++ tests were using the image default compiler -- g++-4.8 for Ubuntu Trusty.

smowton commented:
For some reason had to push this as a branch on `diffblue/cbmc` to get any Travis build running

kroening commented:
May I suggest to split out the extra Travis jobs? I have been restarting Travis jobs all day, and doubt that Travis can handle this.

smowton commented:
Can do in general, but I will leave them in this PR as this change specifically needs CI'ing against gcc-7 and clang-6. Then I'll remove them before merge.

We should move to AWS CodeBuild if Travis can't take this. I read a bit about how we should set this up and messaged @forejtv.

smowton commented:
Rewrote this to extract default language standard at the same time as querying version. Regrettably this means a second run to get `__cplusplus`, which will never be defined at the same time as `__STDC_VERSION__`. Please re-review.

The old idea (derive default standard from compiler version) was undone by Apple claiming Clang 8.1.0, but it is derived from LLVM 3.9.0 and uses default C++98. I could special case that as well... but something somewhere would lie in a different way, so I though it better to ask the actual question I intend, "what language standard do you think you're compiling?"

kroening commented:
I like the new way better, as it relieves us from tracking the default of the various compilers.
Still doubt that Travis can deliver what you expect it to do.

smowton commented:
The build / test matrix here btw is:

Gcc-5 tested with gcc-5
Gcc-7 tested with gcc-7
Clang-3.7 tested with gcc-5
Clang-6.0 tested with gcc-5
Apple clang 8.1.0 tested with self

The non-Apple Clangs are not running tests using the same compiler because they fail tests at the moment. Gcc-8 is not tested as cbmc doesn't build due to a bad catch-by-value (PR coming up) and small_map breaking rules for use of memmove (contacted Dan Poetzel about this)

smowton committed:
Make it clearer that <=> will always be stringified

clang++-6.0 warns that in the next C++ standard <=> will be an operator and therefore an atomic
lexical token. I think that's overly zealous, but it's easily worked around by explicitly quoting
the name, which will be stringified before use in any case.

See https://github.com/diffblue/cbmc/issues/1974 for more details, or
https://bugs.llvm.org/show_bug.cgi?id=36925 for an upstream bug report.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 4491b4af9cdac38c783efe9670fe5e745f85ca10).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76783517

smowton commented:
@kroening looks like this is working now. Travis is coping -- of course this raises the number of Travis jobs in the second stage from 6 to 8. @tautschnig @peterschrammel please review

tautschnig commented on .travis.yml:
```diff
@@ -212,14 +222,74 @@ jobs:
+            - libc6-dev-i386
+      before_install:
+        - mkdir bin
+        # Use gcc/g++ 5 for tests, as Clang doesn't work yet
```
What is it that's going wrong - could you be more verbose about this, please?

tautschnig replied:
Could you please include a reference to #2370 in the comment? (Though we should of course really fix #2370...)

smowton replied:
Already done, rotate your eyeballs down one line ;)

tautschnig replied:
Somehow GitHub should get smarter about this. (Surely it can't be me who is lacking the smarts :-P)

tautschnig commented on src/util/irep_ids.def:
```diff
@@ -41,7 +41,7 @@ IREP_ID_ONE(member_name)
 IREP_ID_TWO(equal, =)
 IREP_ID_TWO(implies, =>)
-IREP_ID_TWO(iff, <=>)
+IREP_ID_TWO(iff, "<=>")
```
See #1974 - I'd say the commit message should include a pointer to that issue.

tautschnig reviewed (approved):
Looks good to me - just minor notes.

smowton commented:
@tautschnig done, see https://github.com/diffblue/cbmc/issues/2370 for specific failures when the native compiler is Clang

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 72101368c03d8edbf37dddf78c891a7bfeb0d3e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76800670
